### PR TITLE
mgr/cephadm: Enhance AlertManagerSpec to allow adding additional webhook receiver URLs

### PR DESCRIPTION
**Pull request by votdev**
Fixes: https://tracker.ceph.com/issues/46775

Signed-off-by: Volker Theile <vtheile@suse.com>

## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/pybind/mgr/cephadm/services/monitoring.py [hunk 1]
```diff
@@ -73,17 +75,25 @@ def config_dashboard(self, daemon_descrs: List[DaemonDescription]):
...
     def create(self, daemon_spec: CephadmDaemonSpec) -> str:
         assert self.TYPE == daemon_spec.daemon_type
+        assert daemon_spec.spec
         return self.mgr._create_daemon(daemon_spec)
 
-    def generate_config(self, daemon_spec: CephadmDaemonSpec) -> Tuple[Dict[str, Any], List[str]]:
+    def generate_config(self, daemon_spec: CephadmDaemonSpec[AlertManagerSpec]) -> Tuple[Dict[str, Any], List[str]]:
         assert self.TYPE == daemon_spec.daemon_type
-        deps = [] # type: List[str]
+        deps = []  # type: List[str]
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/services/monitoring.py:90]:
```suggestion
        deps: List[str] = []
```

cephadm is not getting backported to py2 :tada:

**Comment by sebastian-philipp** [comment]
```
2020-07-31T22:21:29.258 INFO:teuthology.orchestra.run.smithi045.stderr:Error EINVAL: Traceback (most recent call last):
2020-07-31T22:21:29.258 INFO:teuthology.orchestra.run.smithi045.stderr:  File "/usr/share/ceph/mgr/mgr_module.py", line 1171, in _handle_command
2020-07-31T22:21:29.259 INFO:teuthology.orchestra.run.smithi045.stderr:    return self.handle_command(inbuf, cmd)
2020-07-31T22:21:29.259 INFO:teuthology.orchestra.run.smithi045.stderr:  File "/usr/share/ceph/mgr/orchestrator/_interface.py", line 138, in handle_command
2020-07-31T22:21:29.259 INFO:teuthology.orchestra.run.smithi045.stderr:    return dispatch[cmd['prefix']].call(self, cmd, inbuf)
2020-07-31T22:21:29.259 INFO:teuthology.orchestra.run.smithi045.stderr:  File "/usr/share/ceph/mgr/mgr_module.py", line 311, in call
2020-07-31T22:21:29.261 INFO:teuthology.orchestra.run.smithi045.stderr:    return self.func(mgr, **kwargs)
2020-07-31T22:21:29.261 INFO:teuthology.orchestra.run.smithi045.stderr:  File "/usr/share/ceph/mgr/orchestrator/_interface.py", line 100, in <lambda>
2020-07-31T22:21:29.262 INFO:teuthology.orchestra.run.smithi045.stderr:    wrapper_copy = lambda *l_args, **l_kwargs: wrapper(*l_args, **l_kwargs)
2020-07-31T22:21:29.262 INFO:teuthology.orchestra.run.smithi045.stderr:  File "/usr/share/ceph/mgr/orchestrator/_interface.py", line 89, in wrapper
2020-07-31T22:21:29.262 INFO:teuthology.orchestra.run.smithi045.stderr:    return func(*args, **kwargs)
2020-07-31T22:21:29.263 INFO:teuthology.orchestra.run.smithi045.stderr:  File "/usr/share/ceph/mgr/orchestrator/module.py", line 1165, in _apply_misc
2020-07-31T22:21:29.263 INFO:teuthology.orchestra.run.smithi045.stderr:    specs = [ServiceSpec(service_type, placement=placementspec, unmanaged=unmanaged, preview_only=dry_run)]
2020-07-31T22:21:29.263 INFO:teuthology.orchestra.run.smithi045.stderr:TypeError: __init__() got an unexpected keyword argument 'preview_only'
2020-07-31T22:21:29.266 INFO:teuthology.orchestra.run.smithi045.stderr:
```
https://pulpito.ceph.com/swagner-2020-07-31_22:03:23-rados:cephadm-wip-swagner-testing-2020-07-31-1340-distro-basic-smithi/5273869

File: src/python-common/ceph/deployment/service_spec.py [hunk 2]
```diff
@@ -737,3 +738,36 @@ def validate(self):
...
 yaml.add_representer(IscsiServiceSpec, ServiceSpec.yaml_representer)
+
+
+class AlertManagerSpec(ServiceSpec):
+    def __init__(self,
+                 service_type: str = 'alertmanager',
+                 service_id: Optional[str] = None,
+                 placement: Optional[PlacementSpec] = None,
+                 unmanaged: bool = False,
+                 user_data: Optional[Dict[str, Any]] = None,
```
Comment (jschmid1) [inline src/python-common/ceph/deployment/service_spec.py:749]:
```suggestion
                 user_data: Optional[Dict[str, Any]] = None,
                 preview_only: bool = False
```

File: src/python-common/ceph/deployment/service_spec.py [hunk 3]
```diff
@@ -737,3 +738,36 @@ def validate(self):
...
+                 ):
+        assert service_type == 'alertmanager'
+        super(AlertManagerSpec, self).__init__(
+            'alertmanager', service_id=service_id,
+            placement=placement, unmanaged=unmanaged)
```
Comment (jschmid1) [inline src/python-common/ceph/deployment/service_spec.py:754]:
```suggestion
            placement=placement, unmanaged=unmanaged,
            preview_only, preview_only)
```

**Comment by votdev** [comment]
> ```
> 2020-07-31T22:21:29.258 INFO:teuthology.orchestra.run.smithi045.stderr:Error EINVAL: Traceback (most recent call last):
> 2020-07-31T22:21:29.258 INFO:teuthology.orchestra.run.smithi045.stderr:  File "/usr/share/ceph/mgr/mgr_module.py", line 1171, in _handle_command
> 2020-07-31T22:21:29.259 INFO:teuthology.orchestra.run.smithi045.stderr:    return self.handle_command(inbuf, cmd)
> 2020-07-31T22:21:29.259 INFO:teuthology.orchestra.run.smithi045.stderr:  File "/usr/share/ceph/mgr/orchestrator/_interface.py", line 138, in handle_command
> 2020-07-31T22:21:29.259 INFO:teuthology.orchestra.run.smithi045.stderr:    return dispatch[cmd['prefix']].call(self, cmd, inbuf)
> 2020-07-31T22:21:29.259 INFO:teuthology.orchestra.run.smithi045.stderr:  File "/usr/share/ceph/mgr/mgr_module.py", line 311, in call
> 2020-07-31T22:21:29.261 INFO:teuthology.orchestra.run.smithi045.stderr:    return self.func(mgr, **kwargs)
> 2020-07-31T22:21:29.261 INFO:teuthology.orchestra.run.smithi045.stderr:  File "/usr/share/ceph/mgr/orchestrator/_interface.py", line 100, in <lambda>
> 2020-07-31T22:21:29.262 INFO:teuthology.orchestra.run.smithi045.stderr:    wrapper_copy = lambda *l_args, **l_kwargs: wrapper(*l_args, **l_kwargs)
> 2020-07-31T22:21:29.262 INFO:teuthology.orchestra.run.smithi045.stderr:  File "/usr/share/ceph/mgr/orchestrator/_interface.py", line 89, in wrapper
> 2020-07-31T22:21:29.262 INFO:teuthology.orchestra.run.smithi045.stderr:    return func(*args, **kwargs)
> 2020-07-31T22:21:29.263 INFO:teuthology.orchestra.run.smithi045.stderr:  File "/usr/share/ceph/mgr/orchestrator/module.py", line 1165, in _apply_misc
> 2020-07-31T22:21:29.263 INFO:teuthology.orchestra.run.smithi045.stderr:    specs = [ServiceSpec(service_type, placement=placementspec, unmanaged=unmanaged, preview_only=dry_run)]
> 2020-07-31T22:21:29.263 INFO:teuthology.orchestra.run.smithi045.stderr:TypeError: __init__() got an unexpected keyword argument 'preview_only'
> 2020-07-31T22:21:29.266 INFO:teuthology.orchestra.run.smithi045.stderr:
> ```
> 
> [pulpito.ceph.com/swagner-2020-07-31_22:03:23-rados:cephadm-wip-swagner-testing-2020-07-31-1340-distro-basic-smithi/5273869](https://pulpito.ceph.com/swagner-2020-07-31_22:03:23-rados:cephadm-wip-swagner-testing-2020-07-31-1340-distro-basic-smithi/5273869)

The changes seem to be merged into master after the PR as been created. The PR has been adapted now.

**Comment by sebastian-philipp** [comment]
green: https://pulpito.ceph.com/swagner-2020-08-04_10:27:26-rados:cephadm-wip-swagner3-testing-2020-08-04-1030-distro-basic-smithi/

**Pull request merged**